src/core/thread/router_table.hpp

```diff
@@ -350,7 +410,21 @@ class RouterTable : public InstanceLocator
      */
     void ProcessTimerTick(void);
 
+    IteratorBuilder Iterate(void) { return IteratorBuilder(GetInstance()); }
+
 private:
+    class IteratorBuilder : public InstanceLocator
+    {
+    public:
+        IteratorBuilder(Instance &aInstance)
+            : InstanceLocator(aInstance)
+        {
+        }
+
+        Iterator begin(void) { return Iterator(GetInstance()); }
+        Iterator end(void) { return Iterator(GetInstance(), Iterator::kEndIterator); }
```

Comment:
I intentionally avoided this pattern. Two reasons:

- First, having `Iterate()` method allows us to later add different kinds of filtering (e.g. iterate over subset of router entries in certain state(s), ...)
- Second, to make `for (Router &router : Get<RouterTable>())` to work, we would need to publicly add `begin()` and `end()` methods to the `RouterTable()` class. The model used in the PR (i.e., adding a `private` `IteratorBuilder`) keeps these methods hidden, while allowing them to used for range-based `for` loop. Also note that this does not add any code-size overhead. This model is taking advantage of the C++ feature that allows methods on a private nested type within a class to be called while the nested type itself remaining private and hidden.

One side-note with regards to `auto` I would suggest that we limit its use overall (maybe allow it for very specific situations where it helps). For example, between `for (Router &router : ...)` and `for (auto &router : ...)`, the first one is more explicit and reads much cleaner (reader of the code does not need to figure out what the type is).

Reply:
Agree that allowing iteration with a filter (as @abtink suggested ) is more concise and consistent with existing code. 

One issue I have lately is that when I iterate ip6 addresses of a Child, I have to copy each address to a local variable, which is less efficient than ideal. For example, I wanted to iterate only multicast addresses, but all addresses including MlEID are copied. If I want to implement a iteration over a subset of Child addresses without copying other addresses, which pattern can better support it?

Reply:
@abtink  The idea of iterating over a filter makes sense 👍 

>One side-note with regards to auto I would suggest that we limit its use overall (maybe allow it for very specific situations where it helps). For example, between for (Router &router : ...) and for (auto &router : ...), the first one is more explicit and reads much cleaner (reader of the code does not need to figure out what the type is).

Sorry, I got used to write `for (auto &...)`, I agree we should limit usage of `auto` and didn't mean to suggest `auto` at here.

Reply:
> One issue I have lately is that when I iterate ip6 addresses of a Child, I have to copy each address to a local variable, which is less efficient than ideal. For example, I wanted to iterate only multicast addresses, but all addresses including MlEID were copied. If we want to implement a iteration over a subset of Child addresses without copying other addresses, which pattern can better support it?

@simonlingoogle , the reason for copying (as you probably noticed) is that we don't save the full mesh-local address and we just save the IID portion and we would need to create it in address instance provided by caller. The optimization for saving only IID of mesh-local address was mainly to reduce RAM, (since we can have many children any reduction in info per child,  can add up).

I have been planning to simplify the `Child:GetNextIp6Address` and related `Iterator` code. Will think of this (and how to add a filtering model so we can get only multicast or unicast addresses).